src/auth/AuthSessionHandler.cc

```diff
@@ -240,12 +240,234 @@ class AES128CBC_HMACSHA256_StreamHandler : public AuthStreamHandler {
+class AES128GCM_StreamHandler : public AuthStreamHandler {
+  CephContext* const cct;
+  std::unique_ptr<EVP_CIPHER_CTX, decltype(&::EVP_CIPHER_CTX_free)> ectx;
+
+  static constexpr const std::size_t AES_KEY_LEN{16};
+  static constexpr const std::size_t STREAM_AES_IV_LEN{16};
+  static constexpr const std::size_t STREAM_GCM_TAG_LEN{16};
+  static constexpr const std::size_t AES_BLOCK_LEN{16};
+
+  std::array<char, AES_KEY_LEN> connection_secret;
+
+  // using GCC's "Tetra Integer" mode here
+  ceph::uint128_t nonce;
+  static_assert(sizeof(nonce) == STREAM_AES_IV_LEN);
+
+public:
+  AES128GCM_StreamHandler(CephContext* const cct,
+			  const AuthConnectionMeta& auth_meta,
+			  const ceph::uint128_t& nonce)
+    : cct(cct),
+      ectx(EVP_CIPHER_CTX_new(), EVP_CIPHER_CTX_free),
+      nonce(nonce) {
+    ceph_assert_always(auth_meta.connection_secret.length() >= AES_KEY_LEN);
+    ceph_assert_always(ectx);
+
+    std::copy_n(std::cbegin(auth_meta.connection_secret), AES_KEY_LEN,
+		std::begin(connection_secret));
+    //EVP_CIPHER_CTX_reset(ectx.get());
+  }
+
+  ~AES128GCM_StreamHandler() override {
+    memset(&nonce, 0, sizeof(nonce));
+    //connection_secret.zero();
+  }
+
+  std::size_t calculate_padded_size(const std::size_t length) {
+    // we need to take into account the PKCS#7 padding. There *always* will
+    // be at least one byte of padding. This stays even to input aligned to
+    // AES_BLOCK_LEN. Otherwise we would face ambiguities during decryption.
+    // To exemplify:
+    //   length := 10 -> p2align(10, 16) + 16 == 16 (06 bytes for padding),
+    //   length := 16 -> p2align(16, 16) + 16 == 32 (16 bytes for padding).
+    //
+    // Technically padding will be added by OpenSSL but its format and even
+    // presence is a part of the public interface, I believe.
+    return p2align(length, AES_BLOCK_LEN) + AES_BLOCK_LEN;
+  }
+
+  std::size_t calculate_payload_size(const std::size_t length) override {
+    // as we're doing *authenticated encryption* additional space is needed
+    // to store Auth Tag. OpenSSL defaults this to 96 bits (12 bytes).
+    const std::size_t authenticated_and_encrypted_size = \
+      calculate_padded_size(length) + STREAM_GCM_TAG_LEN;
+    return authenticated_and_encrypted_size;
+  }
+
+  void authenticated_encrypt(ceph::bufferlist& payload) override;
+  void authenticated_decrypt(char* payload, uint32_t& length) override;
+};
+
+void AES128GCM_StreamHandler::authenticated_encrypt(ceph::bufferlist& payload)
+{
+  if (1 != EVP_EncryptInit_ex(ectx.get(), EVP_aes_128_gcm(),
+			      nullptr, nullptr, nullptr)) {
+    throw std::runtime_error("EVP_EncryptInit_ex failed");
+  }
+
+  if (1 != EVP_CIPHER_CTX_ctrl(ectx.get(), EVP_CTRL_GCM_SET_IVLEN,
+			       STREAM_AES_IV_LEN, nullptr) ) {
+    throw std::runtime_error("EVP_CIPHER_CTX_ctrl failed");
+  }
+
+  if(1 != EVP_EncryptInit_ex(ectx.get(), nullptr, nullptr,
+	reinterpret_cast<const unsigned char*>(connection_secret.data()),
```

Comment:
Hmm, the internals suggest we don't need to reinitialize key each time (AES key setup was pretty expensive). IV can be set separately: 

```c
static int aesni_gcm_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
                              const unsigned char *iv, int enc)
{
    EVP_AES_GCM_CTX *gctx = ctx->cipher_data;
    if (!iv && !key)
        return 1;
    if (key) {
        aesni_set_encrypt_key(key, ctx->key_len * 8, &gctx->ks.ks);
        CRYPTO_gcm128_init(&gctx->gcm, &gctx->ks, (block128_f) aesni_encrypt);
        gctx->ctr = (ctr128_f) aesni_ctr32_encrypt_blocks;
        /*
         * If we have an iv can set it directly, otherwise use saved IV.
         */ 
        if (iv == NULL && gctx->iv_set)
            iv = gctx->iv; 
        if (iv) {
            CRYPTO_gcm128_setiv(&gctx->gcm, iv, gctx->ivlen);
            gctx->iv_set = 1;
        }
        gctx->key_set = 1;
    } else {
        /* If key set use IV, otherwise copy */
        if (gctx->key_set)
            CRYPTO_gcm128_setiv(&gctx->gcm, iv, gctx->ivlen);
        else
            memcpy(gctx->iv, iv, gctx->ivlen);
        gctx->iv_set = 1;
        gctx->iv_gen = 0;
    }
    return 1;
}
```